Persistence API for Akka Typed, #22273

patriknw opened the pull request:
Here is a first stab at a persistence implementation for Akka Typed.

It also includes several examples to illustrate the API, by @raboof .

It's running the Behavior as an untyped PersistentActor to be able to use the difficult implementation from untyped `EventSourced` trait. This has a major limitation. The behavior can't be wrapped in things like `Actor.deferred`, `Actor.supervise`, or `Actor.withTimers`. Because of that limitation this will not be the final approach. I think we have to re-implement the needed parts of `EventSourced` in a real typed `Behavior`. For this sprint, as a prototype, it should be good enough though.

Refs #22273

raboof committed:
Some experiments with a typed api for persistence

* Expand example with a timer
* Move 'onRecoveryComplete' to 'builder' style
* rather than using a default parameter which is hard to
  evolve with binary compatibility
* Explicit snapshots
* Example spawning a child actor
* Taking a snapshot needs access to the state
* Added signal handler
* I like how simple the types of `commandHandler` and
 `signalHandler` are now, but it'd be nice if they were closer together and could perhaps even share the `state` switch. I'll give that a try.
* Attempt to get signal handling closer to command handling
* Always use the full state for snapshots
* If you want to add a mapping layer use event adapters or a custom
  serializer.
* Example of an actor that separates 'persistent' and transient state
* Use ordering 'command, event, state, context' consistently
* Allow multiple `andThen`
* Replace ask with spawned adapters
* Allow 'andThen' on each PersistentEffect
* Make 'onSignal' a partial function
* Remove unnecessary parens
* Simplify 'andThen', make sure it receives the State

patriknw committed:
first impl of Typed PersistentActor, #22273

* illustrate problem with wrapping behavior

raboof committed:
Move ApiTest to PersistentActorCompileOnlyTest

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+    persistenceId: String,
+    initialState:  State,
+    actions:       Actions[Command, Event, State],
+    onEvent:       (Event, State) ⇒ State): PersistentBehavior[Command, Event, State] =
```
I think I would prefer to name `onEvent` to `applyEvent`.
`onEvent` sounds like the purpose is for side effects.

ktoso replied:
Sounds good 👍

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+  }
+
+  type CommandHandler[Command, Event, State] = Function3[Command, State, ActorContext[Command], PersistentEffect[Event, State]]
```
I realized that the order of the parameters might need another round of discussion. In other typed APIs we have the context as the first parameter. My thinking was that context was least interesting when reading. Having it differently is not an option, so the question is if we just place it first here or change all other places?

raboof replied:
I think "command, event, state, context" and "context, state, event, command" are both fine (leave out components as necessary :) ).

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+    new PersistentBehavior(persistenceId, initialState, actions, onEvent,
+      recoveryCompleted = (state, _) ⇒ state)
+
+  sealed abstract class PersistentEffect[+Event, State]() {
```
this should probably be named `Effect`?

raboof replied:
Works for me, the package gives a hint anyway

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+    def andThen(callback: State ⇒ Unit): PersistentEffect[Event, State]
+  }
+
+  final case class PersistNothing[Event, State](callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
```
I probably want to rewrite the callbacks impl in the way we did at the other place. More optimized for the 0 and 1 case. Also, now they are executed in "wrong" order.

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+  }
+
+  case class Unhandled[Event, State](callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
```
rename to `UnhandledCommand` ?

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+
+  def snapshotOnState(predicate: State ⇒ Boolean): PersistentBehavior[Command, Event, State] = ??? // FIXME
+
+  def snapshotOn(predicate: (State, Event) ⇒ Boolean): PersistentBehavior[Command, Event, State] = ??? // FIXME
```
snapshots are not implemented yet

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/internal/PersistentActorImpl.scala:
```diff
@@ -0,0 +1,114 @@
+
+    case msg ⇒
+      try {
+        // FIXME sigHandler(state)
```
this FIXME is done, but should be tested

patriknw committed:
impl and test of signals

raboof commented on akka-typed-tests/src/test/scala/akka/typed/persistence/scaladsl/PersistentActorCompileOnlyTest.scala:
```diff
@@ -0,0 +1,381 @@
+
+      actions = Actions((cmd, state, ctx) ⇒ cmd match {
+        case DoSideEffect(data) ⇒
+          Persist[Event, EventsInFlight](IntentRecorded(state.nextCorrelationId, data)).andThen { _ ⇒
```
I'm not too happy with requiring the `[Event, EventsInFlight]` type parameters here (and the compiler errors are really confusing if you get them wrong or leave them out). I wonder if making `ctx` a `PersistentActorContext[Event, State]` and adding a `def thenPersist(...) = Persist[Event, State](...)` would be an improvement, I'll give that a try.

raboof commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+      recoveryCompleted = (state, _) ⇒ state)
+
+  sealed abstract class PersistentEffect[+Event, State]() {
+    def andThen(callback: State ⇒ Unit): PersistentEffect[Event, State]
```
Should we also add an `andThenStop`? Or perhaps this is the time to try introducing a `Stop` effect and a way to combine effects?

patriknw replied:
right, both approaches would probably work, I'll try something

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/persistence/scaladsl/PersistentActorCompileOnlyTest.scala:
```diff
@@ -0,0 +1,381 @@
+    PersistentActor.persistent[RegisterTask, Event, State](
+      persistenceId = "asdf",
+      initialState = State(Nil),
+      // The 'onSignal' seems to break type inference here.. not sure if that can be avoided?
```
yes, I think we have the same "problem" with `Actor.immutable` + `.onSignal`

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/internal/PersistentActorImpl.scala:
```diff
@@ -0,0 +1,121 @@
+            actions.sigHandler(state).applyOrElse((sig, state, ctx), unhandledSignal)
+          case a.ReceiveTimeout ⇒
+            actions.commandHandler(ctxAdapter.receiveTimeoutMsg, state, ctx)
+          // TODO note that PostStop and PreRestart signals are not handled, we wouldn't be able to persist there
```
@raboof I implemented the signals, but realized that currently the only signal that makes sense is `Terminated`. Perhaps we should have a more specific `onTerminated` handler instead? Or we skip signals altogether, and recommend `watchWith`.

Note that `ReceiveTimeout` is not a signal.

raboof committed:
Chaining side effects (callbacks and stopping)

Cleans up the API in a couple of places, but still needs type parameters
in spots where I wouldn't want to need them...

Allow multiple side effects in one 'andThen'

Making type inference actually work

Allow applying multiple side effects

In a way that allows 'inspecting' the effects if you want to, but
doesn't require you to repeat generic parameters when you don't.

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+
+object PersistentActor {
+  def persistent[Command, Event, State](
+    persistenceId: String,
```
this will have to be `ActorContext => String` to be able to work with `ClusterSharding`, since the `persistenceId` is not known when creating the Behavior

patriknw committed:
Some additional api adjustments, #22273

* rename PersistentEffect to Effect
* change order of parameters, ctx first
* rename onEvent to applyEvent
* persistenceIdFromActorName for Cluster Sharding
* PersistenActor.immutable

patriknw commented:
@raboof I have fixed the remaining small things. I think this should be good to be merged now. Let me know when you have looked and I'll squash some of my commits here (leaving your's unchanged).

patriknw commented:
rebased, will merge after validation

patriknw commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,140 @@
+import akka.typed.scaladsl.ActorContext
+
+object PersistentActor {
+  def persistent[Command, Event, State](
```
I like the idea of `PersistentActor.immutable`, who knows, we might have a mutable one also, at least for java (where type inference will completely break down).

ktoso replied:
Yeah, exactly since the immutable style there gets really boilerplaty... Shall we ticketify?

ktoso replied:
Did it https://github.com/akka/akka/issues/23693

dispalt commented on akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala:
```diff
@@ -0,0 +1,190 @@
+    new PersistentBehavior(persistenceIdFromActorName, initialState, actions, applyEvent,
+      recoveryCompleted = (_, state) ⇒ state)
+
+  sealed abstract class Effect[+Event, State]() {
```
I kinda feel like there should be an escape hatch for setting the State directly.  The inverse of SideEffect.

raboof replied:
Thanks for looking into this! If you have part of the state persistent and managed via events, but also want part of the state to be modifiable directly from commands, our current thinking is that  using `Actor.deferred` and a `var` for the directly-modifiable state like in https://github.com/akka/akka/pull/23674/files#diff-be682d29ed550fd2c88b3ce6e8eeed26R297 might be a clean way to do it. What do you think?

patriknw replied:
yes, changing the state without an event might be dangerous/surprising. If it's replayed (restarted) it will not have the same state, and that can matter when applying events that were persisted after "patching" the state.

dispalt replied:
Actually, and sorry this might take a bit to explain, is as follows.  I will say I totally understand if you don't want to support this usecase because its a little convoluted. I also understand that while I have good behavior around this, it may encourage bad behavior.

I usually have multiple persists in response to a single command.  Right now I have a runtime (a shim over untyped), that basically handles flatMap'ing over multiple persists and getting the state immediately as if the event was persisted.  And I can build up some nice, sophisticated workflows with it.  So my little mini programs run and output `(S, Seq[E], ...)`.  So in this case it would be nice to have that exposed.  But I can work around it by just duplicating the effort.  But the case for escape hatching would be that I do understand how it can effect the recovery, but I have good hygiene so maybe an escape hatch would be nice.

dispalt replied:
@raboof also yes, that mechanism is how I keep things like actorRef's (that are transient).  Thanks for the discussion.
